src/wp-includes/functions.php

```diff
@@ -8458,3 +8458,87 @@ function wp_recursive_ksort( &$array ) {
 	}
 	ksort( $array );
 }
+
+/**
+ * Attempt to detect a VirtualBox environment.
+ *
+ * This attempts all known methods of detecting VirtualBox.
+ *
+ * @since 6.1.0
+ *
+ * @global WP_Filesystem_Base $wp_filesystem WordPress filesystem subclass.
+ *
+ * @return bool Whether or not VirtualBox was detected.
+ */
+function is_virtualbox() {
+	global $wp_filesystem;
+	static $is_virtualbox;
+
+	if ( null !== $is_virtualbox ) {
+		return $is_virtualbox;
+	}
+
+	/**
+	 * Filters whether the current environment uses VirtualBox.
+	 *
+	 * @since 6.1.0
+	 *
+	 * @param bool Whether the current environment uses VirtualBox. Default false.
+	 */
+	if ( apply_filters( 'is_virtualbox', false ) ) {
+		$is_virtualbox = true;
+		return $is_virtualbox;
+	}
+
+	// Detection via Composer.
+	if ( function_exists( 'getenv' ) && 'virtualbox' === getenv( 'COMPOSER_RUNTIME_ENV' ) ) {
+		$is_virtualbox = true;
+		return $is_virtualbox;
+	}
```

Comment:
I think this is a poorer choice than a dedicated WP environment variable/constant. It now relies on a constant intended for third party software than a minority of sites use.

The most popular WP packages, [johnpbloch/wordpress](https://packagist.org/packages/johnpbloch/wordpress) and [roots/wordpress](https://packagist.org/packages/roots/wordpress), have around 208K installs combined in the last 30 days; WP has [20M+ for 6.0](https://wordpress.org/download/counter/) alone.

Reply:
To clarify, this isn't relying on the existence of a third party environment variable to determine _whether_ it's a VirtualBox environment. It's taking advantage of an already existing configuration that says "This should be treated as a VirtualBox environment".

In the event that this environment variable is not set, the function continues to attempt other detection methods. It doesn't assume that its absence means it's not VirtualBox.

This would mean that Core wouldn't have to commit to the BC promise for a new, WP-specific environment variable, which Core committers disagreed with.

The meaning of setting the Composer environment variable to `virtualbox` isn't going to change. It will always explicitly state that the application should consider this to be a VirtualBox environment - no other usage for it makes sense. In that sense, I think the BC-promise is safe.

Aside from Composer, this was also influenced by [Drupal's detection methods](https://api.drupal.org/api/drupal/vendor%21composer%21composer%21src%21Composer%21Util%21Platform.php/function/Platform%3A%3AisVirtualBoxGuest/9.3.x).

Reply:
Testing for VirtualBox has been removed in favor of those environments providing specific fixes for the issue. As far as core/2225 is concerned, adding the pre_move_dir and post_move_dir hooks is sufficient to enable this type of VB-specific fix now and for any potential similar issue in the future.